onnx/defs/loss/defs.cc

```diff
@@ -0,0 +1,146 @@
+            FunctionBodyHelper::BuildNodes({
+                // nodes: {outputs, op, inputs, attributes}
+		{{"X_SM"}, "Softmax", {"scores"}},
+		{{"X_Log"}, "Log", {"X_SM"}},
+                {{"output"}, "Mul", {"labels", "X_Log"}}
+		}))
+	.AddQueriedFunctionBody([](FunctionBodyQueryContext& ctx) { // no weight, reduction is "sum"
+              return ctx.getNumInputs() == 2 && ctx.getAttribute("reduction")->s() == "sum"; },
+            FunctionBodyHelper::BuildNodes({
+                // nodes: {outputs, op, inputs, attributes}
+                {{"X_SM"}, "Softmax", {"scores"}},
+                {{"X_Log"}, "Log", {"X_SM"}},
+                {{"X_Mul"}, "Mul", {"labels", "X_Log"}},
+                {{"output"}, "ReduceSum", {"X_Mul"}}
+                }))
+	.AddQueriedFunctionBody([](FunctionBodyQueryContext& ctx) { // no weight, reduction is "mean"
+              return ctx.getNumInputs() == 2 && ctx.getAttribute("reduction")->s() == "mean"; },
+            FunctionBodyHelper::BuildNodes({
+                // nodes: {outputs, op, inputs, attributes}
+                {{"X_SM"}, "Softmax", {"scores"}},
+                {{"X_Log"}, "Log", {"X_SM"}},
+                {{"X_Mul"}, "Mul", {"labels", "X_Log"}},
+                {{"output"}, "ReduceMean", {"X_Mul"}}
+                }))
+	.AddQueriedFunctionBody([](FunctionBodyQueryContext& ctx) { // weight, reduction is "none"
+              return ctx.getNumInputs() > 2 && ctx.getAttribute("reduction")->s() == "none"; },
+            FunctionBodyHelper::BuildNodes({
+                // nodes: {outputs, op, inputs, attributes}
+                {{"X_SM"}, "Softmax", {"scores"}},
+                {{"X_Log"}, "Log", {"X_SM"}},
+                {{"X_Mul"}, "Mul", {"labels", "X_Log"}},
+                {{"output"}, "Mul", {"weights", "X_Mul"}}
+                }))
+        .AddQueriedFunctionBody([](FunctionBodyQueryContext& ctx) { // weight, reduction is "sum"
+              return ctx.getNumInputs() > 2 && ctx.getAttribute("reduction")->s() == "sum"; },
+            FunctionBodyHelper::BuildNodes({
+                // nodes: {outputs, op, inputs, attributes}
+                {{"X_SM"}, "Softmax", {"scores"}},
+                {{"X_Log"}, "Log", {"X_SM"}},
+                {{"X_Mul"}, "Mul", {"labels", "X_Log"}},
+		{"X_Mul2"}, "Mul", {"weights", "X_Mul"}},
+                {{"output"}, "ReduceSum", {"X_Mul2"}}
+                }))
+        .AddQueriedFunctionBody([](FunctionBodyQueryContext& ctx) { // no weight, reduction is "mean"
+              return ctx.getNumInputs() > 2 && ctx.getAttribute("reduction")->s() == "mean"; },
+            FunctionBodyHelper::BuildNodes({
+                // nodes: {outputs, op, inputs, attributes}
+                {{"X_SM"}, "Softmax", {"scores"}},
+                {{"X_Log"}, "Log", {"X_SM"}},
+                {{"X_Mul"}, "Mul", {"labels", "X_Log"}},
+		{{"X_Mul2"}, "Mul", {"weights", "X_Mul"}},
+                {{"output"}, "ReduceMean", {"X_Mul2"}}
```

Comment:
>ReduceMean [](start = 30, length = 10)

do we need to set axes here ? 
Does default value works ?

Reply:
no need to specify axes here.

If the reduction is specified, ReduceMean will be computed over the whole ouput (axes=None, and output will be scalar).
For more specific cases requiring computing the mean over specific axes, there is the possibility of using a subgraph with ONNX::SoftmaxCrossEntropy with 'none' reduction followed ONNX::ReduceMean with the desired axes.

How does that sound?